fusion/fusion/FusionGroup.hpp

```diff
@@ -0,0 +1,303 @@
+    };
+
+    // ---------------------------------------------------------------
+    //! \brief FusionGroup parent
+    //!
+    //! opcs & asm statements are not supported yet, future
+    // ---------------------------------------------------------------
+    class FusionGroupBase
+    {
+      public:
+        //! \brief save typing
+        using UidType = fusion::UidType;
+        //! \brief save typing
+        using HashType = fusion::HashType;
+        //! \brief save typing
+        using InstPtrListType = fusion::InstPtrListType;
+        //! \brief save typing
+        using InstUidListType = fusion::InstUidListType;
+
+        //! \brief base ctor
+        FusionGroupBase(std::string n = "", InstUidListType u = InstUidListType(), HashType h = 0) :
+            name_(n),
+            uids_(u),
+            // Future feature
+            // opcs_(fusion::OpcodeListType()),
+            // asms_(fusion::AsmStmtListType()),
+            hash_(h)
+        {
+        }
+
+        //! \brief capture the UIDs and create the hash key
+        virtual void setUids(InstUidListType & u)
+        {
+            uids_ = u;
+            initHash();
+        }
+
+        //! \brief uids accessor
+        virtual InstUidListType & uids() { return uids_; }
+
+        // virtual OpcodeListType&  opcs()     { return opcs_; }
+        // virtual AsmStmtListType& asms()     { return asms_; }
+
+        //! \brief hash setter
+        virtual void setHash(HashType hash) { hash_ = hash; }
+
+        //! \brief refresh the hash from uids_
+        virtual void initHash() { hash_ = jenkins_1aat(uids_); }
+
+        //! \brief hash accessor
+        virtual HashType hash() const { return hash_; }
+
+        //! \brief convenience function
+        virtual void setName(std::string n) { name_ = n; }
+
+        //! \brief name accessor
+        virtual std::string name() const { return name_; }
+
+        //! \brief report fgroup state to stream
+        //!
+        //! I prefer this instead of overloading <<
+        virtual void info(std::ostream & os = std::cout) const
+        {
+            std::cout << "Name: " << name() << std::endl;
+            std::cout << "  HASH: " << std::hex << "  0x" << hash() << std::endl;
+            std::cout << "  UIDS: ";
+            for (auto u : uids_)
+                std::cout << std::hex << "  0x" << u;
+            std::cout << std::endl;
+            // Future feature
+            // std::cout<<"  OPCs: ";
+            // for(auto o : opcs_) std::cout<<std::hex<<"  0x"<<o;
+            // std::cout<<std::endl;
+            // std::cout<<"  ASM: ";
+            // for(auto a : asms_) std::cout<<a<<std::endl;
+        }
+
+      protected:
+        //! \brief Method to calculate hash based on UIDs
+        //!
+        //! In future consider making this a user controlled functor.
+        //! In hardware the adds are not desirable.
+        static HashType jenkins_1aat(const std::vector<UidType> & v)
```

Comment:
Realize this may get replaced with a user-controlled functor in the future and be moot, but was curious if it would be worth adding a comment as to the logic behind the implemented hash function.  Maybe similar to the comments regarding the RadixTrie (which I appreciated!).

Reply:
I have added some additional description.